src/gui/gui.cpp

```diff
@@ -38,34 +38,94 @@ void init()
 	try {
-		schema_validation::schema_validator validator(filesystem::get_wml_location("schema/gui.cfg").value());
+		schema_validation::schema_validator validator(filesystem::get_wml_location(schema_file).value());
 
-		preproc_map preproc(game_config::config_cache::instance().get_preproc_map());
-		filesystem::scoped_istream stream = preprocess_file(filesystem::get_wml_location("gui/_main.cfg").value(), &preproc);
+		// Core theme files
+		current_file = "gui/_main.cfg";
+		filesystem::scoped_istream stream = preprocess_file(filesystem::get_wml_location(current_file).value(), &preproc);
 		read(guis_cfg, *stream, &validator);
+
 	} catch(const config::error& e) {
 		ERR_GUI_P << e.what();
-		ERR_GUI_P << "Setting: could not read file 'data/gui/_main.cfg'.";
+		ERR_GUI_P << "Setting: could not read gui file: " << current_file;
 	} catch(const abstract_validator::error& e) {
-		ERR_GUI_P << "Setting: could not read file 'data/schema/gui.cfg'.";
+		ERR_GUI_P << "Setting: could not read schema file: " << schema_file;
 		ERR_GUI_P << e.message;
 	}
 
 	//
-	// Parse GUI definitions.
+	// Read and validate theme WML files from addons
+	//
+
+	// Add the $user_campaign_dir/*/gui.cfg files to the addon gui config.
+	std::vector<std::string> user_dirs;
+	{
+		const std::string user_campaign_dir = filesystem::get_addons_dir();
+		std::vector<std::string> user_files;
+		filesystem::get_files_in_dir(
+			user_campaign_dir, &user_files, &user_dirs, filesystem::name_mode::ENTIRE_FILE_PATH);
+	}
+
+	for(const std::string& umc : user_dirs) {
+		try {
+			const std::string gui_file = umc + "/gui-theme.cfg";
+			current_file = filesystem::get_short_wml_path(gui_file);
+			if(filesystem::file_exists(gui_file)) {
+				config addon_cfg;
+				schema_validation::schema_validator validator(filesystem::get_wml_location(schema_file).value());
+				read(addon_cfg, *preprocess_file(gui_file, &preproc), &validator);
+				addons_cfg.append(addon_cfg);
+			}
+		} catch(const config::error& e) {
+			ERR_GUI_P << e.what();
+			ERR_GUI_P << "Setting: could not read gui file: " << current_file;
+		} catch(const abstract_validator::error& e) {
+			ERR_GUI_P << "Setting: could not read schema file: " << schema_file;
+			ERR_GUI_P << e.message;
+		}
+	}
+
+	//
+	// Parse GUI definitions from mainline
 	//
 	for(const config& g : guis_cfg.child_range("gui")) {
 		const std::string id = g["id"];
 
-		auto iter = guis.emplace(id, gui_definition(g)).first;
+		auto [iter, is_unique] = guis.emplace(id, gui_definition(g));
 
-		if(id == "default") {
-			default_gui = iter;
+		if (!is_unique) {
+			ERR_GUI_P << "GUI Theme ID '" << id << "' already exists.";
+		} else {
+			if(id == "default") {
+				default_gui = iter;
+			}
+		}
+	}
+
+	//
+	// Parse GUI definitions from addons
+	//
+	for(const config& g : addons_cfg.child_range("gui")) {
+		const std::string id = g["id"];
+
+		try {
+			auto [iter, is_unique] = guis.emplace(id, gui_definition(g));
```

Comment:
Use 
```cpp
auto [iter, is_unique] = guis.try_emplace(id, g);
```
to avoid creating unnecessary `gui_definition` objects.

Reply:
Updated.